### PR TITLE
revised circular queue challenge and added working tests

### LINKED ISSUE #13428: [beta] How to Create a Circular Queue : Instructions need to be clearer
Refers changes from : https://github.com/freeCodeCamp/freeCodeCamp/pull/13308

The instructions sections ned to be more clear, step by step as the tests are currently. The introduction part is clear. But the instructions need to be step to step.

### PR DESCRIPTION
**Pull request by bonham000**
#### Pre-Submission Checklist

- [x] Your pull request targets the `staging` branch of freeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] All new and existing tests pass the command `npm test`. Use `git commit --amend` to amend any fixes.

#### Type of Change

- [x] Small bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

#### Checklist:

- [x] Tested changes locally.
- [x] Closes #13428

#### Description
This PR revises the Circular Queue challenge in the Advanced Data Structures section with a improved description, instructions, seed code, and working tests.

### CODE REVIEW

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json [hunk 1]
```diff
@@ -192,9 +192,9 @@
     },
     {
       "id": "587d8255367417b2b2512c75",
-      "title": "How to Create a Circular Queue",
+      "title": "Create a Circular Queue",
       "description": [
-        "In this challenge you will be creating a Circular Queue. A circular queue is basically a queue that writes to the end of a collection then begins over writing itself at the beginning of the collection. This is type of data structure has some useful applications in certain situations. For example, a circular buffer can be used in streaming media. Once the queue is full, new media data simply begins to overwrite old data.",
+        "In this challenge you will be creating a Circular Queue. A circular queue is basically a queue that writes to the end of a collection then begins over writing itself at the beginning of the collection. This is type of data structure has some useful applications in certain situations. For example, a circular buffer can be used to stream media. Once the queue is full, new media data simply begins to overwrite old data.",
```
Comment (Greenheart) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: I'd say "a circular queue is commonly used for streaming media" might be a better wording. That way, we're clearer about the use case, and avoid introducing a related word (but not the precise one we refer to in this challenge). What do you think? :blush:

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json [hunk 2]
```diff
@@ -203,35 +203,54 @@
...
       ],
       "type": "waypoint",
       "releasedOn": "Feb 17, 2017",
-      "solutions": [
-        "class CircularQueue {\n  constructor () {\n    this._cq = ['', '', '', '', ''];\n    this._reader = 0;\n    this._writer = 0;\n  }\n  \n  printCollection () {\n    return this._cq;\n  }\n  \n  _updatePointer (id) {\n    if (++this[id] === this._cq.length) {\n      this[id] = 0;\n    }\n  }\n  \n  enqueue (item) {\n    if (Array.isArray(item)) {\n      for (const x of item) {\n        this.enqueue(x);\n      }\n\n      return;\n    }\n    \n    this._cq[this._writer] = item;\n    \n    this._updatePointer('_writer');\n  }\n  \n  dequeue () {\n    const item = this._cq[this._reader];\n    this._cq[this._reader] = '';\n    \n    this._updatePointer('_reader');\n    return item;\n  }\n  \n  front () {\n    return this._cq[this._reader];\n  }\n}"
```
Comment (Greenheart) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: Why did you delete the solution? It's used to improve the coverage of `npm test`, which allows us to see when challenges get breaking tests etc

Comment (Greenheart) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: Ah, I created this solution for the old version of the challenge, so it's no longer valuable.

I think we should add an updated solution.

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json [hunk 3]
```diff
@@ -203,35 +203,54 @@
...
+        "In this challenge we will implement a circular queue. The circular queue should provide <code>enqueue</code> and <code>dequeue</code> methods which allow you to read and write to the queue. The class itself should also accept an integer which you can use to specify the size of the queue when you create it. We've written the starting version of this class for you in the code editor. When you enqueue items to the queue, the write pointer should advance forward and loop back to the beginning once it reaches the end of the queue. Likewise, the read pointer should advance forward as you dequeue items. The write pointer should not be allowed to move past the read pointer (our class won't let you overwrite data you haven't read yet) and the read pointer should not be able to advance past data you have written.",
```
Comment (Greenheart) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: What do you think?
```diff
-read and write to the queue
+read from and write to the queue
```

**Review by Greenheart — CHANGES_REQUESTED** [review]
@bonham000 Thanks for working on this! :smile:

**Comment by bonham000** [comment]
@Greenheart hey I made the two revisions and I didn't realize `npm test` was verifying solution code if it's present in the seed files which is the real reason I removed that from the original PR. Anyway, I added in my solution now and verified it with `npm test` so this should be good to go now.

File: seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json [hunk 4]
```diff
@@ -192,9 +192,9 @@
     },
     {
       "id": "587d8255367417b2b2512c75",
-      "title": "How to Create a Circular Queue",
+      "title": "Create a Circular Queue",
       "description": [
-        "In this challenge you will be creating a Circular Queue. A circular queue is basically a queue that writes to the end of a collection then begins over writing itself at the beginning of the collection. This is type of data structure has some useful applications in certain situations. For example, a circular buffer can be used in streaming media. Once the queue is full, new media data simply begins to overwrite old data.",
+        "In this challenge you will be creating a Circular Queue. A circular queue is basically a queue that writes to the end of a collection then begins over writing itself at the beginning of the collection. This is type of data structure has some useful applications in certain situations. For example, a circular queue can be used to stream media. Once the queue is full, new media data simply begins to overwrite old data.",
```
Comment (Greenheart) [inline seed/challenges/08-coding-interview-questions-and-take-home-assignments/coding-interview-data-structure-questions.json]: If I'm not confusing this for something else, the circular queue technically isn't responsible for the *streaming of media* as the current sentence indicates - it's useful when storing the streaming media.

I'd suggest the following change
```diff
-to stream media
+for streaming media
```

**Review by Greenheart — CHANGES_REQUESTED** [review]
@bonham000 Nice solution! Just one more thing :smile:

**Comment by bonham000** [comment]
@Greenheart sounds good, I've added that change!

**Review by Greenheart — APPROVED** [review]
@bonham000 Thanks! This looks good :blush:

**Pull request merged**